cpp/src/decisiontree/batched-levelalgo/builder_base.cuh

```diff
@@ -0,0 +1,479 @@
+    CUDA_CHECK(cudaMemsetAsync(n_depth, 0, sizeof(IdxT), s));
+    if (isRegression()) {
+      CUDA_CHECK(
+        cudaMemsetAsync(block_sync, 0, sizeof(char) * block_sync_size, s));
+    }
+    node_start = 0;
+    node_end = h_total_nodes = 1;  // start with root node
+    h_nodes[0].initSpNode();
+    h_nodes[0].start = 0;
+    h_nodes[0].count = input.nSampledRows;
+    h_nodes[0].depth = 0;
+  }
+
+  /** check whether any more nodes need to be processed or not */
+  bool isOver() const { return node_end == h_total_nodes; }
+
+  /**
+   * @brief After the current batch is finished processing, update the range
+   *        of nodes to be worked upon in the next batch
+   */
+  void updateNodeRange() {
+    node_start = node_end;
+    auto nodes_remaining = h_total_nodes - node_end;
+    node_end = std::min(nodes_remaining, params.max_batch_size) + node_end;
+  }
+
+  /**
+   * @brief Computes best split across all nodes in the current batch and splits
+   *        the nodes accordingly
+   *
+   * @param[out] h_nodes list of nodes (must be allocated using cudaMallocHost!)
+   * @param[in]  s cuda stream
+   * @return the number of newly created nodes
+   */
+  IdxT doSplit(NodeT* h_nodes, cudaStream_t s) {
+    auto batchSize = node_end - node_start;
+    // start fresh on the number of *new* nodes created in this batch
+    CUDA_CHECK(cudaMemsetAsync(n_nodes, 0, sizeof(IdxT), s));
+    initSplit<DataT, IdxT, Traits::TPB_DEFAULT>(splits, batchSize, s);
+    // get the current set of nodes to be worked upon
+    MLCommon::updateDevice(curr_nodes, h_nodes + node_start, batchSize, s);
+    // iterate through a batch of columns (to reduce the memory pressure) and
+    // compute the best split at the end
+    auto n_col_blks = n_blks_for_cols;
+    for (IdxT c = 0; c < input.nSampledCols; c += n_col_blks) {
+      Traits::computeSplit(*this, c, batchSize, params.split_criterion, s);
+      CUDA_CHECK(cudaGetLastError());
+    }
+    // create child nodes (or make the current ones leaf)
+    auto smemSize = Traits::nodeSplitSmemSize(*this);
+    nodeSplitKernel<DataT, LabelT, IdxT, typename Traits::DevTraits,
+                    Traits::TPB_SPLIT>
+      <<<batchSize, Traits::TPB_SPLIT, smemSize, s>>>(
+        params.max_depth, params.min_rows_per_node, params.max_leaves,
+        params.min_impurity_decrease, input, curr_nodes, next_nodes, n_nodes,
+        splits, n_leaves, h_total_nodes, n_depth);
+    CUDA_CHECK(cudaGetLastError());
+    // copy the updated (due to leaf creation) and newly created child nodes
+    MLCommon::updateHost(h_nodes + node_start, curr_nodes, batchSize, s);
+    MLCommon::updateHost(h_n_nodes, n_nodes, 1, s);
+    CUDA_CHECK(cudaStreamSynchronize(s));
+    MLCommon::updateHost(h_nodes + h_total_nodes, next_nodes, *h_n_nodes, s);
+    return *h_n_nodes;
+  }
+};  // end Builder
+
+/**
+ * @brief Traits used to customize the Builder for classification task
```

Comment:
Nitpick: Why a Trait? I think the term "policy class" is more commonly used in the C++ ecosystem.

Reply:
I'll let @teju85 take this one.